parsons/action_builder/action_builder.py

```diff
@@ -0,0 +1,351 @@
+
+    def add_tags_to_record(
+        self, identifiers, tag_name, tag_field, tag_section, campaign=None
+    ):
+        """
+        Add a tag (i.e. custom field value) to an existing entity record in Action Builder. The
+        tags, along with their category/field and section/group, must already exist (except for
+        date fields). If fewer fields are provided than tags, the excess tags will be applied to
+        the final field provided. Similarly, the final section will be applied to any fields or
+        tags in excess.
+        `Args:`
+            identifiers: list
+                The list of strings of unique identifiers for a record being updated. ID strings
+                will need to begin with the origin system, followed by a colon, e.g.
+                `action_builder:abc123-...`.
+            tag_name: str or list
+                The name(s) of the new tag(s), i.e. the custom field value(s).
+            tag_field: str or list
+                The name(s) of the tag category(ies), i.e. the custom field name(s). Must have
+                equal or fewer items as `tag_name`.
+            tag_section: str or list
+                The name(s) of the tag section(s), i.e. the custom field group name(s). Must have
+                equal or fewer items as `tag_section`.
+            campaign: str
+                Optional. The 36-character "interact ID" of the campaign whose data is to be
+                retrieved or edited. Not necessary if supplied when instantiating the class.
+        `Returns:`
+            Dict containing Action Builder entity data of the entity being tagged.
+        """
+
+        # Ensure all tag args are lists
+        tag_name = tag_name if isinstance(tag_name, list) else [tag_name]
+        tag_field = tag_field if isinstance(tag_field, list) else [tag_field]
+        tag_section = tag_section if isinstance(tag_section, list) else [tag_section]
+
+        # Use lists of tuples to identify length ordering in case amounts differ
+        lengths = []
+        lengths.append(
+            ("name", len(tag_name))
+        )  # Most specific - should be >= other lengths
+        lengths.append(("field", len(tag_field)))
+        lengths.append(
+            ("section", len(tag_section))
+        )  # Most generic - should be <= other lengths
+
+        # Sort by length, maintaining original order in ties
+        ordered_lengths = sorted(lengths, key=lambda x: x[1], reverse=True)
+        sorted_keys = [x[0] for x in ordered_lengths]
+
+        # Raise an error if there are fewer specific items provided than generic
+        if sorted_keys[0] != "name":
+            raise ValueError(
+                "Not enough tag_names provided for tag_fields or tag_sections"
+            )
+
+        if sorted_keys[1] != "field":
+            raise ValueError("Not enough tag_fields provided for tag_sections")
+
+        # Construct tag data, applying final generic taxonomy to any excess specific items
```

Comment:
I'm not sure I'm following but this behavior doesn't seem intuitive to me and could lead to some surprising data.

Reply:
As above, if there isn't a way to make this more intuitive than requiring a list of dicts, I'm not sure this method provides any added convenience.